### DISCUSSION TITLE
Drawing a multicolored triangle

### DISCUSSION THREAD

**Discussion by CarlSandstrom** [vsg-dev/VulkanSceneGraph#730] [Q&A]
I am trying to re-create the classic example of a multicolored triangle in VulkanSceneGraph.

<img src="https://user-images.githubusercontent.com/3179136/221178464-518818ae-aa7e-4e0b-85fb-ff77b88a5338.png" width="350">

To do this, I use a StateGroup object produced by the vsg::Builder::createStateGroup() method as the root node. Furthermore, I create vertices, normals, texcoords, colors and indices as follows:

```
vsg::ref_ptr<vsg::vec3Array> vertices = vsg::vec3Array::create({vsg::vec3(0, 0, 0), vsg::vec3(1, 0, 0), vsg::vec3(0, 1, 0)});
vsg::ref_ptr<vsg::vec3Array> normals = vsg::vec3Array::create({vsg::vec3(0, 0, 1), vsg::vec3(0, 0, 1), vsg::vec3(0, 0, 1)});
vsg::ref_ptr<vsg::vec2Array> texcoords = vsg::vec2Array::create({vsg::vec2(0, 0), vsg::vec2(0, 0), vsg::vec2(0, 0)});
vsg::ref_ptr<vsg::vec4Array> colors = vsg::vec4Array::create({vsg::vec4(1, 0, 0, 1), vsg::vec4(0, 1, 0, 1), vsg::vec4(0, 0, 1, 1)});
vsg::ref_ptr<vsg::ushortArray> indices = vsg::ushortArray::create({0, 1, 2});
```

Then I use the vsg::VertexIndexDraw class and add it to the StateGroup.

```
auto vid = vsg::VertexIndexDraw::create();
vsg::DataList arrays;

arrays.push_back(vertices);
arrays.push_back(normals);
arrays.push_back(texcoords);
arrays.push_back(colors);

vid->assignArrays(arrays);
vid->assignIndices(indices);
vid->indexCount = static_cast<uint32_t>(indices->size());
vid->instanceCount = 1;

stateGroup->addChild(vid);
```

The program draws the triangle as expected, but the color is of the triangle is red. I have also tried using the vsg::Commands class but the result is a black triangle, even if a add a light source.

What am I missing? Is it wrong to pass a vector of vectors as color?

**Comment by robertosfield** [comment]
Hi Carl,

It's hard to pinpoint what might be going on without having the source
code, but my best guess would be that the state set up is using a per
instance rather than per vertex binding for the colour array.  The
vsg::Builder class is built around one color per geometry instances so the
binding are set up for this.

It would be more complicated to do more of the state setup yourself but
will give you full control, so for now I'd recommend against using
vsg::Builder a usage case it wan't designed for.  Longer term I plan to
expand vsg::Builder to better support freeform scene graph setup rather
than primitive shapes that it does right now.

To roll your own geometry you could take a low level take as show in the
vsgdraw example, or use a high level approach like used by vsg::Builder.
For the later the vsggraphicspipelineconfiguratior is the place to start,
in this example it uses a per instance colour, but

https://github.com/vsg-dev/vsgExamples/blob/master/examples/utils/vsggraphicspipelineconfigurator/vsggraphicspipelineconfigurator.cpp#L142

To testing things ghanging this to a color array setup to a vsg::vec4Array
with the same number of entries are there are vertices then change the line:

           graphicsPipelineConfig->assignArray(vertexArrays, "vsg_Color",
VK_VERTEX_INPUT_RATE_INSTANCE, colors);

To:
           graphicsPipelineConfig->assignArray(vertexArrays, "vsg_Color",
VK_VERTEX_INPUT_RATE_VERTEX, colors);

Then when the graphicsPipelineConfig->init() is invoked it'll create a
graphics pipeline that exacts per vertex colours.

I hope this helps,
Robert.

**Reply by CarlSandstrom** [reply]
I followed the vsggraphicspipelineconfigurator example with your suggested modification and it worked like a charm :-)

Thank you and have a nice weekend!
